test/e2e/targets/awss3/main.go

```diff
@@ -0,0 +1,364 @@
+	. "github.com/onsi/gomega"    //nolint:stylecheck
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+
+	"github.com/triggermesh/triggermesh/test/e2e/framework"
+	"github.com/triggermesh/triggermesh/test/e2e/framework/apps"
+	e2es3 "github.com/triggermesh/triggermesh/test/e2e/framework/aws/s3"
+	e2ece "github.com/triggermesh/triggermesh/test/e2e/framework/cloudevents"
+	"github.com/triggermesh/triggermesh/test/e2e/framework/ducktypes"
+)
+
+/* This test suite requires:
+
+   - AWS credentials in whichever form (https://docs.aws.amazon.com/sdk-for-go/api/aws/session/#hdr-Sessions_options_from_Shared_Config)
+   - The name of an AWS region exported in the environment as AWS_REGION
+*/
+
+var targetAPIVersion = schema.GroupVersion{
+	Group:   "targets.triggermesh.io",
+	Version: "v1alpha1",
+}
+
+const (
+	targetKind     = "AWSS3Target"
+	targetResource = "awss3targets"
+)
+
+const awsAccessKeyIDSecretKey = "awsApiKey"
+const awsSecretAccessKeySecretKey = "awsApiSecret"
+
+var _ = Describe("AWS S3 target", func() {
+	f := framework.New("awss3target")
+	region := os.Getenv("AWS_REGION")
+
+	var ns string
+
+	var trgtClient dynamic.ResourceInterface
+
+	var bucketARN string
+	var bucketName string
+	var awsCreds credentials.Value
+	var awsSecret *corev1.Secret
+
+	BeforeEach(func() {
+		ns = f.UniqueName
+
+		gvr := targetAPIVersion.WithResource(targetResource)
+		trgtClient = f.DynamicClient.Resource(gvr).Namespace(ns)
+	})
+
+	Context("a target is deployed", func() {
+		var trgtURL *url.URL
+		var s3Client *s3.S3
+
+		BeforeEach(func() {
+			sess := session.Must(session.NewSession())
+			s3Client = s3.New(sess)
+			awsCreds = readAWSCredentials(sess)
+
+			By("creating a S3 bucket", func() {
+				bucketName = e2es3.CreateBucket(s3Client, f, region)
+				bucketARN = createBucketARN(bucketName, region)
+				awsSecret = createAWSCredsSecret(f.KubeClient, ns, awsCreds)
+			})
+		})
+
+		AfterEach(func() {
+			By("deleting S3 bucket "+bucketName, func() {
+				e2es3.DeleteBucket(s3Client, bucketName)
+			})
+		})
+
+		When("the spec contains default settings", func() {
+			BeforeEach(func() {
+				By("creating an AWSS3Target object", func() {
+					trgt, err := createTarget(trgtClient, ns, "test-",
+						withARN(bucketARN),
+						withCredentials(awsSecret.Name),
+					)
+					Expect(err).ToNot(HaveOccurred())
+
+					trgt = ducktypes.WaitUntilReady(f.DynamicClient, trgt)
+
+					trgtURL = ducktypes.Address(trgt)
+					Expect(trgtURL).ToNot(BeNil())
+				})
+			})
+
+			When("an event is sent to the target", func() {
+				var sentEvent *cloudevents.Event
```

Comment:
Instead of declaring this variable twice (here, and in the sibling Context), it could be moved under `Context("a target is deployed"`